### ISSUE TITLE
[thing actions] Actions fail to execute after a binding restart

### ISSUE THREAD

**Issue by mhilbush** [openhab/openhab-core#1265]
Reference my comment [here](https://github.com/openhab/openhab2-addons/pull/6494#issuecomment-561128456).

Most (if not all) thing action implementations contain code that looks like this.
```
    public static void restart(@Nullable ThingActions actions) {
        if (actions instanceof DoorbirdActions) {
            ((DoorbirdActions) actions).restart();
        } else {
            throw new IllegalArgumentException("Instance is not a DoorbirdActionsService class");
        }
    }
```

Normally, the `instanceof` test succeeds. However, after installing a new version of the binding, this test fails, which causes the code to throw `IllegalArgumentException` when invoking the action from a rule. This condition can only be cleared by restarting openHAB. There are several ways to reproduce this, one of which is to drop a new version of a binding into `addons` that overwrites an older version that's already there.

After installing the new version of the binding, it appears that the classloader instances are different. I suspect this is what's causing the `instanceof` test to fail.

Before installing new binding version:
```
2019-12-09 09:23:02.018 [INFO ] [ding.doorbird.action.DoorbirdActions] - DoorbirdActions classloader org.eclipse.osgi.internal.loader.EquinoxClassLoader@50578098[org.openhab.binding.doorbird:2.5.0.201912091420(id=245)]
2019-12-09 09:23:02.019 [INFO ] [ding.doorbird.action.DoorbirdActions] - ThingActions actions classloader org.eclipse.osgi.internal.loader.EquinoxClassLoader@50578098[org.openhab.binding.doorbird:2.5.0.201912091420(id=245)]
```
After installing new binding version:
```
2019-12-09 09:24:39.813 [INFO ] [ding.doorbird.action.DoorbirdActions] - DoorbirdActions classloader org.eclipse.osgi.internal.loader.EquinoxClassLoader@50578098[org.openhab.binding.doorbird:2.5.0.201912091423(id=245)]
2019-12-09 09:24:39.814 [INFO ] [ding.doorbird.action.DoorbirdActions] - ThingActions actions classloader org.eclipse.osgi.internal.loader.EquinoxClassLoader@1c14e241[org.openhab.binding.doorbird:2.5.0.201912091423(id=245)]
```

**Comment by cweitkamp** [comment]
Thanks for reporting this issue. I can remember that I faced it back in the days too.

The injection of the `ThingActions` classes into DSL is done [here](https://github.com/openhab/openhab-core/blob/12e8edc03931ded800af688b04d826d2fda0bd2c/bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/scoping/ScriptImplicitlyImportedTypes.java). I am not sure about the lifecycle but it looks like it is only relevant during restart. Adding something during runtime is not taken into account.

**Comment by kaikreuzer** [comment]
Afaik, it works dynamically at runtime for the `ActionService` instances, though (when installing an "Action add-on", it becomes available in rules without a restart).
There should thus be some hidden mechanism that re-initializes the scripts/rules upon changes - will will have to find it and make sure that it also considers `ThingActions`.

**Comment by mhilbush** [comment]
Will this issue also impact the NGRE and, if so, would it also need to be fixed there?

**Comment by kaikreuzer** [comment]
No.

**Comment by kaikreuzer** [comment]
> No.

Looking closer into it, I am not that sure anymore. The issue might be the registration of the service through the BaseThingHandler, although I would expect this to be disposed and newly created... But your logs appear as if it is this instance, which is still the old one after the update, which is weird. If that's the case, I'd assume that the NGRE could be affected as well.

It looks complex and I am afraid, I won't be able to solve it before the release. 😞

**Comment by mhilbush** [comment]
Thanks for looking into this. To be clear, my log statements were placed immediately in front of the `if (actions instanceof DoorbirdActions) {` statement in the above code.

> It looks complex and I am afraid, I won't be able to solve it before the release. 

Totally makes sense. I also thought it might be a bit too risky to fix before the release.

**Comment by mhilbush** [comment]
FWIW, the problem can be worked around by replacing this
```
public static void restart(@Nullable ThingActions actions) {
    if (actions instanceof DoorbirdActions) {
        ((DoorbirdActions) actions).restart();
    } else {
        throw new IllegalArgumentException("Instance is not a DoorbirdActionsService class");
    }
}
```

With this
```
public static void restart(@Nullable ThingActions actions) {
    if (actions == null) {
        throw new IllegalArgumentException("actions cannot be null");
    }
    try {
        actions.getClass().getDeclaredMethod("restart").invoke(actions));
    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
        logger.info("Exception invoking method", e);
    }
}
```

@kaikreuzer @cweitkamp IDK if you would consider this to be an acceptable solution, or whether you'd prefer to solve the root cause (although I have no idea how it might be solved...). It is a more fragile solution, and it gets a little messy when there are quite a few arguments to the action.

**Comment by cweitkamp** [comment]
@mhilbush Thanks for your investigation. I can confirm your suggested solution is working (minor syntax changes needed). But you are right. Your suggested solution will be very clumsy if you have arguments or return values.

I digged deeper into this topic. It took some time because I had to setup a test environment for me (in this case I am using a new OH3 Pushover binding which is a port from OH1 binding I am currently working on).

The registration (and removal) of the `ThingActions` services is working properly. I added some debugging to relevant locations and observed this:

Starting the system:
```
10:46:04.614 [le-processing-3] INFO  o.o.c.a.m.s.i.d.DefaultScriptScopeProvider:168 - DefaultScriptScopeProvider: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:46:04.623 [le-processing-3] INFO  o.o.c.model.script.ScriptServiceUtil:137 - ScriptServiceUtil: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:46:04.624 [le-processing-3] INFO  o.o.c.m.s.i.e.a.ThingActionService:93 - ThingActionService: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
```

Refreshing of binding:
```
10:48:06.675 [safeCall-2] INFO  o.o.c.a.m.s.i.d.DefaultScriptScopeProvider:176 - DefaultScriptScopeProvider: Removed ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:48:06.675 [safeCall-2] INFO  o.o.c.model.script.ScriptServiceUtil:143 - ScriptServiceUtil: Removed ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:48:06.675 [safeCall-2] INFO  o.o.c.m.s.i.e.a.ThingActionService:101 - ThingActionService: Removed ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'

10:48:06.795 [pool-1-thread-1] INFO  o.o.c.a.m.s.i.d.DefaultScriptScopeProvider:168 - DefaultScriptScopeProvider: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:48:06.796 [pool-1-thread-1] INFO  o.o.c.model.script.ScriptServiceUtil:137 - ScriptServiceUtil: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
10:48:06.796 [pool-1-thread-1] INFO  o.o.c.m.s.i.e.a.ThingActionService:93 - ThingActionService: Added ThingAction 'org.openhab.binding.pushover.internal.actions.PushoverThingActions'
```

But I was able to tackle down the issue to be related to the use of the `ActionClassLoader`. IIRC this classloader instance is bound to the XTEXT rule engine and responsible for loading `ActionService`s and `ThingActions`. Its 'loadClass()' method is called only once - the first time one will try to access `ThingActions` from a XTEXT rule. I guess the result will be cached somewhere. @kaikreuzer Any ideas?

**Comment by mhilbush** [comment]
@cweitkamp Thanks for chasing this down. I guess that explains why it was/is not an issue with the classes that extend `ActionService`, as those classes seem to contain no static methods?

**Comment by cweitkamp** [comment]
Compared to the `ActionService`s there is one difference: There is a `RulesRefresher` component (e.g. `RulesThingRefresher`) which triggers a refresh of the DSL rules whenever a `ActionService` is added. It does not handle `ThingActions` services. Maybe that is the place to look into.

**Comment by mhilbush** [comment]
I saw one of these today. I definitely didn't install a new version of the Mail binding, so there must be something else that caused the error. 
```
2020-01-29 19:00:00.012 [ERROR] [ntime.internal.engine.ExecuteRuleJob] - Error during the execution of rule 'Battery Health Monitor': Instance is not of class SendMailActions.
```
I suppose the problem will occur if something causes the binding to go through a dispose/initialize cycle. I did install a new binding several hours before the above problem occurred. Does that cause all other bindings to dispose/initialize? If so, then I would suggest this issue is more severe than I originally thought, as the all thing actions will break whenever a binding is installed (or uninstalled).

**Comment by lolodomo** [comment]
Any news ? Because this bug is terribly annoying. 
Should we implement the workaround in each binding?

**Comment by kaikreuzer** [comment]
I would wait what refactorings will be done wrt dropping the old rule engine and just keeping the NGRE. The situation will likely change a lot, so I don't think we should spend much effort right now.

**Comment by lolodomo** [comment]
The problem is that it is difficult to keep this issue during one or two years...

**Comment by kaikreuzer** [comment]
Not sure what you mean. OH 3 is planned in a few months, not years.

**Comment by lolodomo** [comment]
How many? At least 9. That is a lot when you face of such issue.

**Comment by mhilbush** [comment]
Providing an update on this...

@cpmeister came up with a workaround for me that I used on the Ecobee binding. It's pretty simple and works well.

The last thing you want is for you to lose the ability to control your thermostat(s) just because you installed or uninstalled some other binding...

https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/action/EcobeeActions.java#L89

**Comment by kaikreuzer** [comment]
Nice that you have found a workaround, but we should indeed find a better solution than implementing that in every single binding...
With a bit of luck, https://github.com/openhab/openhab-core/pull/1451 might make a difference, we will see.

**Comment by mhilbush** [comment]
Agreed. It's an easy workaround, and it's easily removed.

It wasn't clear to me how #1451 would affect this issue (but admittedly I don't understand this area very well). Also, I don't think we ever definitively proved whether the NGRE also was impacted by this issue. Would you like me to do that to establish a baseline?

**Comment by kaikreuzer** [comment]
It isn't yet clear to me either, I was just mentioning that it might make a difference...

How would you currently test it in the NGRE? Within a JS/Python script? Sure, would be interesting how the thing actions behave there.

**Comment by mhilbush** [comment]
Yep. Would use a Python script.

**Comment by mhilbush** [comment]
The good news is that so far I've not been able to replicate the issue using the NGRE.

**Comment by cweitkamp** [comment]
Played a little bit with creating `ThingActions`. Noticed an issue when restarting a binding multiple times (or removing / replacing a `*.jar` file in the addons-folder). Not sure if it is related but I have the slight feeling that it is:

The options inside a config description of existing actions are duplicated each time. On the first run I only could select one Thing for the action. after a restart there were two options, another restart there were three, and so on:

![grafik](https://user-images.githubusercontent.com/5131747/86126602-dfe42b80-bade-11ea-8d36-e0ca1828c6b5.png)

![grafik](https://user-images.githubusercontent.com/5131747/86127124-b2e44880-badf-11ea-9711-c52cf960fec7.png)

There obviously is a bug inside creating / refreshing `ThingActions` procedure.

**Comment by cweitkamp** [comment]
I found the culprit. The `ModuleTypeProvider` for `AnnotatedActions` should consider the scope (by default `ActionScope`) when parsing the `ModuleType`s. The `ThingActions` define their own scope: `ThingActionsScope`. It is the implementation uses it for adding but not for removing actions - this leads to unrelated references in memory:

https://github.com/openhab/openhab-core/blob/6c935985a62321218828e37a9c4a302c628358f5/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/thingsupport/AnnotatedThingActionModuleTypeProvider.java#L71-L75

https://github.com/openhab/openhab-core/blob/6c935985a62321218828e37a9c4a302c628358f5/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/thingsupport/AnnotatedThingActionModuleTypeProvider.java#L107-L108

**Comment by lolodomo** [comment]
We habe now in 18 OH2 bindings with ThingActions feature.
I see at least 3 of them (ecobee, enigma2 and lcn) already solved the current issue with their method `invokeMethodOf`.
I propose we fix the other bindings by using the same method.

**Comment by kaikreuzer** [comment]
I would very much hope that we can fix it for openHAB 3 - so all of those PRs should be considered as a temporary fix only, which we should revert for openHAB 3 again.

**Comment by druciak** [comment]
Why can't we make this workaround generic and put it into the core (or other common library) until we have a better solution? In the bindings there would be just a call to such generic static method and that's it.
It does not make sense to duplicate the same method body in every binding.

**Comment by J-N-K** [comment]
Because core is frozen.

**Comment by druciak** [comment]
Yeah, that's right. I thought it is frozen for new features, not bugfixes.

**Comment by Hilbrand** [comment]
What is blocking us from making a bug fix release for core? These ugly workarounds are lots of work and also are not very inviting to add more thing actions to current bindings?

**Comment by lolodomo** [comment]
13 of 18 bindings are already fixed. 
3 others are waiting for a merge of the fix. 
2 are not yet fixed.

**Comment by kaikreuzer** [comment]
@Hilbrand All 2.5.x patch releases sind 2.5.0 release are patching the add-ons only, but keep using 2.5.0 core bundles. We do not have any build and release processes in place for delivering fixes in core. Furthermore, the addons are automatically updating without the need update the distro at all - so those installations wouldn't benefit.
Efforts are currently rather on moving to master/3.x than to revive 2.5.x for core bundles...

**Comment by lolodomo** [comment]
18 of 19 bindings are fixed in the release 2.5.7.
Unfortunately, the fix I proposed for the xmppclient binding was not merged.
